docs/source/common/optimizers.rst

```diff
@@ -67,27 +67,91 @@ Here is the same example as above using a ``closure``.
             opt.zero_grad()
 
 
-.. code-block:: python
-
-    # Scenario for a GAN.
-    def training_step(...):
-        opt_gen, opt_dis = self.optimizers()
+.. tip:: Be careful where you call ``zero_grad`` or your model won't converge. It is good pratice to call ``zero_grad`` before ``manual_backward``.
 
-        # compute generator loss
-        loss_gen = self.compute_generator_loss(...)
-
-        # zero_grad needs to be called before backward
-        opt_gen.zero_grad()
-        self.manual_backward(loss_gen)
-        opt_gen.step()
 
-        # compute discriminator loss
-        loss_dis = self.compute_discriminator_loss(...)
+.. code-block:: python
 
-        # zero_grad needs to be called before backward
-        opt_dis.zero_grad()
-        self.manual_backward(loss_dis)
-        opt_dis.step()
+    import torch
+    from pytorch_lightning import LightningModule
+    from torch.utils.data import Dataset
+
+    class SimpleGAN(LightningModule):
+
+        def __init__(self):
+            super().__init__()
+            latent_dim = 64
+            self._Z = MultivariateNormal(tr.zeros(latent_dim, device=self.device),
+                                        tr.eye(latent_dim, device=self.device))
+            self.G = MnistDenseGenerator(latent_dim)
+            self.D = MnistDenseDiscriminator()
+            self.num_workers = 0
+
+        @property
+        def automatic_optimization(self):
+            # Important: This property activate ``manual optimization`` for this model
+            return False
+
+        def train_dataloader(self) -> DataLoader:
+            return tr.utils.data.DataLoader(data, batch_size=64, shuffle=True,
+                                            num_workers=self.num_workers)
+
+        def forward(self, x):
+            return self.G(x)
+
+    def generator_loss(self, d_z: Tensor) -> Tensor:
+        # the closer ``d_z`` is from 1,
+        # the better the generator is able to fool the discriminator
+        return -1 * tr.log(d_z).mean()
+
+    def discriminator_loss(self, d_x: Tensor, d_z: Tensor) -> Tensor:
+        # the closer is ``d_x`` from 1 and ``dz`` from 0,
+        # the better the discriminator is able to distinguish
+        # true data from generated ones
+        return -1 * (tr.log(d_x).mean() + tr.log(1 - d_z).mean())
+
+    def sample_z(self, n) -> Tensor:
+        sample = self._Z.sample((n,))
+        return sample
+
+    def sample_G(self, n) -> Tensor:
+        z = self.sample_z(n)
+        return self.G(z)
```

Comment:
Is it worth having all this? A reader still wouldnt be able to run it as `MnistDenseGenerator` is not available

Reply:
Good catch ! Removing it.